### ISSUE TITLE
[Feature Request] [Inductor][dynamic_shape]Size-specialized autotune config for non-template (pointwise & reduction) kernels under dynamic shapes

### ISSUE THREAD

**Issue by markc-614** [pytorch/pytorch#197842]
### 🚀 The feature, motivation and pitch

Under dynamic shapes, Inductor autotunes a single launch config for a given kernel and reuses it for every subsequent runtime shape. The chosen config (e.g. block size for pointwise, or split factor / RSPLIT for reduction) is decided by the first shape the kernel is autotuned with. A config that is optimal for a small shape can be catastrophic for a large shape — even though the shape is dynamic and the kernel could have been re-tuned.

This affects non-template triton kernels in general:

- pointwise kernels (block size XBLOCK / YBLOCK / ZBLOCK, num_warps)
- reduction kernels (split factor / RSPLIT, XBLOCK, num_warps)

and is backend-independent (CUDA, MLU, or any SIMD backend).

## Root cause

- CachingAutotuner.run benchmarks once and sets self.launchers = [best]; all later shapes reuse that one config.
- Pointwise/reduction configs are generated from compile-time size_hints, which is a single hint for dynamic shapes.
- Block size / split factor / num_warps are compile-time constants in Triton, so adapting to size requires different configs/binaries rather than a single adaptive binary.

Concrete CUDA (B300, PyTorch version: 2.13.0+cu130) reproductions with torch.sum (reduction) and torch.add (pointwise), using dynamic=True + mode="max-autotune".

### reduction
```
import os
import torch
torch.set_float32_matmul_precision("high")

def red(x):
    return torch.sum(x)

c = torch.compile(red, dynamic=True, mode="max-autotune")

start = torch.cuda.Event(enable_timing=True)
end = torch.cuda.Event(enable_timing=True)

def time_it(fn, x, warmup=5, reps=7):
    for _ in range(warmup):
        fn(x)
    torch.cuda.synchronize()
    ts = []
    for _ in range(reps):
        start.record(); fn(x); end.record(); torch.cuda.synchronize()
        ts.append(start.elapsed_time(end))
    return sorted(ts)[len(ts) // 2]

x_small = torch.randn(128, 128, device="cuda")
print("small, small-tuned:", time_it(c, x_small))

x_large = torch.randn(8192, 8192, device="cuda")
print("large, reusing small-tuned:", time_it(c, x_large))

os.environ["TORCHINDUCTOR_CACHE_DIR"] = f"/tmp/tc_d_{os.getpid()}"
torch._dynamo.reset()

d = torch.compile(red, dynamic=True, mode="max-autotune")
print("large, fresh autotune:", time_it(d, x_large))
```

output
```

small, small-tuned:            0.016 ms
large, reusing small-tuned:   12.576 ms
large, fresh autotune:         0.223 ms
```
-> slow / fast ratio: 56.3x

### pointwise
```
import os
import torch

def binary(x, y):
    return torch.add(x, y)

c = torch.compile(binary, dynamic=True, mode="max-autotune")

start = torch.cuda.Event(enable_timing=True)
end = torch.cuda.Event(enable_timing=True)

def time_it(fn, args, warmup=5, reps=7):
    for _ in range(warmup):
        fn(*args)
    torch.cuda.synchronize()
    ts = []
    for _ in range(reps):
        start.record()
        fn(*args)
        end.record()
        torch.cuda.synchronize()
        ts.append(start.elapsed_time(end))
    return sorted(ts)[len(ts) // 2]

x_small = torch.randn(128, 128, device="cuda")
y_small = torch.randn(128, 128, device="cuda")
print("small, small-tuned:", time_it(c, (x_small, y_small)))

x_large = torch.randn(16384, 16384, device="cuda")
y_large = torch.randn(16384, 16384, device="cuda")
print("large, reusing small-tuned:", time_it(c, (x_large, y_large)))

os.environ["TORCHINDUCTOR_CACHE_DIR"] = f"/tmp/tc_d_{os.getpid()}"
torch._dynamo.reset()

d = torch.compile(binary, dynamic=True, mode="max-autotune")
print("large, fresh autotune:", time_it(d, (x_large, y_large)))
```
output
```
small, small-tuned:            0.019 ms
large, reusing small-tuned:    1.489 ms
large, fresh autotune:         1.142 ms
```
-> slow / fast ratio: ~1.30x

The regression is operator-dependent (dramatic for reduction, milder but consistent for pointwise), but the root cause is the same: a single small-shape-tuned config is frozen and reused for large shapes.

## Desired behavior

Make config selection size-specialized for non-template triton kernels: partition the input-size space into buckets (or generate a small set of size-hinted kernels) and dispatch to the config/kernel appropriate for the observed runtime size — for both pointwise and reduction.

### size-bucketed autotune

A working prototype patches CachingAutotuner._precompile_worker and CachingAutotuner.run to make config selection size-dependent. In brief:

1. At codegen, record each dynamic dim's value range into inductor_meta.
2. At runtime, geometrically sample the size space (growth factor ~4) into buckets, capped at an empirical max (2^20) above which the best block size stops changing.
3. Eagerly precompile one best config per bucket, then dispatch to the bucket whose size is closest to the actual runtime size.

This works and fixes the large-shape slowdown, but has downsides:

- Increased device memory: every per-bucket kernel is compiled and kept resident, so binary/workspace footprint grows with the number of buckets.
- OOM risk during autotune: benchmarking each bucket requires constructing example inputs at that bucket's scale (up to the empirical cap), which can exhaust device memory — especially for large buckets or many buckets.

### Shape-range guards at the dynamo level

Partition the size space into ranges and rely on dynamo guards so that crossing a range boundary triggers a recompile, each specialization autotuning for its own range. Trade-offs:

- P99 latency fluctuation: crossing a range boundary triggers a recompile + re-autotune on the critical path, producing a latency spike.

### Extend multi_kernel_hints / SizeHintMultiKernel to non-template kernels

Reuse the existing size-hint multi-kernel dispatch (compile-time N kernels + runtime log2-L1 distance selection) for pointwise and reduction. This avoids the eager per-bucket precompilation and the resulting memory blow-up, but requires a non-template analog of MultiTemplateBuffer.

### Additional context

1. Does the community already have a plan or an existing feature that addresses size-specialized config selection for non-template (pointwise/reduction) kernels?
2. Which direction is preferred: runtime bucketing inside CachingAutotuner, dynamo shape-range guards, or extending multi_kernel_hints to pointwise/reduction?

If there is already a plan or an in-flight solution in this area, we'd be glad to align with and contribute to it; if not, we'd be happy to help drive one forward.

cc @chauhang @penguinwu @ezyang @bobrenjc93 @aditvenk @laithsakka @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by markc-614** [comment]
@chauhang @penguinwu @ezyang @bobrenjc93  @shunting314 @eellison  @laithsakka @zou3519 @jansel @Chillee - would appreciate your input on this.

**Comment by eellison** [comment]
cc @laithsakka . Yes i agree this would be valuable. More discussion on apis & heuristics of how this fires could be discussed after implementation.
